enhancements/installer/openstack-baremetal-compute.md

```diff
@@ -0,0 +1,311 @@
+---
+title: openstack-baremetal-compute
+authors:
+  - "@tomassedovic"
+  - "@pierreprinetti"
+reviewers:
+  - "@shiftstack/team"
+approvers:
+  - TBD
+creation-date: 2019-12-10
+last-updated: 2019-12-18
+status: provisional
+
+---
+
+# Openstack Baremetal Compute
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OpenStack's [Nova][openstack-nova] service can be configured to provision
+baremetal machines using the same mechanisms as the virtual machines.
+
+This enhancement will allow [Ironic][openstack-ironic]-provisioned machines to
+be used as OpenShift Compute Nodes, when they are made available as Nova
+flavors.
+
+The OpenShift deployment will have a virtualised control plane and a mix of
+virtual and baremetal compute nodes. Both can be managed by the Machine API. By
+virtue of using the same tools and processes as the fully virtualised OpenStack
+deployments, it will provide the same integration with OpenStack services (e.g.
+storage, load balancer).
+
+## Motivation
+
+In heavy-loaded clusters, or when the workload is better handled with specially
+purposed hardware, the functional requirement of controlling where the work is
+computed might outweight the added value of virtualisation.
+
+In OpenStack clusters where Nova transparently provisions baremetal machines,
+this enhancement will ensure that OpenShift installations can use them as
+Compute Nodes.
+
+Existing OpenShift clusters running on a baremetal-enabled OpenStack will be
+able to add baremetal workers by adding a new node pool as a day-two operation.
+
+### Goals
+
+* Enable running OpenShift on a mix of virtual and baremetal machines on an OpenStack tenant.
+* Provide the same integration with the underlying OpenStack services as in the fully virtualised environment.
+* Baremetal machines can be added and removed from a running OpenShift cluster using the Machine API.
+* Support both `flat` and `neutron` Ironic Network Interfaces
+  - /cc Ironic folks: will this complicate matters significantly?
```

Comment:
A network is a network... Neutron networks just have to be vlan tagged if they are able to be mapped to physical machines. A vxlan network, realistically won't be tagged to a physical machine unless there is a provider in the infrastructure performing translation and attachment. The unfortunate aspect is there is no way to discover this externally.

I would advise asking what network and then rolling forward from there, and maybe document the type of network that it needs to be and in what scenarios it will and will not work.

The only other aspect is that if the neutron network interface is to be used, the ironic node link_local_information with LLDP data needs to be populated. But that is not the responsibility of a non-privileged API client. The deployment will fail if the baremetal machine is configured for neutron network_interface, and similarly it is NOT the responsibility or capability of an un-privileged API user to change those driver settings in ironic. An operator has made an explicit decision as to how the baremetal nodes integrate into their infrastructure and configured the infrastructure accordingly. In other words, I'm worried even considering or focusing on these items may become a red herring.

Reply:
Thanks, that makes sense.

I was hoping we could make the implementation simpler if we could just support the `neutron` network interface. But that would limit the usefulness of this feature since a lot of people use the `flat` one. And we will want to support adding a node to an existing network soon enough anyway.